sugar.go

```diff
@@ -212,25 +208,35 @@ func (s *SugaredLogger) sweetenFields(args []interface{}) []zapcore.Field {
 	if len(args) == 0 {
 		return nil
 	}
-	if len(args)%2 == 1 {
-		s.core.DPanic(_oddNumberErrMsg, Any("ignored", args[len(args)-1]))
-	}
 
-	fields := make([]zapcore.Field, len(args)/2)
-	for i := range fields {
-		keyIdx := 2 * i
-		val := args[keyIdx+1]
-		key, ok := args[keyIdx].(string)
+	// Allocate enough space for the common case; if users only pass structured
+	// fields, we'll need to grow the slice.
+	fields := make([]zapcore.Field, 0, len(args)/2)
+	var i int
+	for i < len(args) {
+		if f, ok := args[i].(zapcore.Field); ok {
```

Comment:
Can we make the state machine structure clearer please? Something like:
```
for i < len(args) {
  // check for inlined zap Fields
  if f, ok := ... { ... }

  // parity check: we need an even number of key/val pairs (at least since the last zap Field)
  if i == len(args)-1 { ... }]

  // consume a key/val pair
  key, val := args[i], args[i+1]

  // process valid string key
  if keyString, ok := key.(string); ok {
    fields = append(fields, Any(keyString, val))
    i += 2
   continue
  }

  // we have a "key" that isn't a string, crash in development...
  s.core.DPanic(...)

  // ...but in production, stop field conversion here, since this could be a parity error (e.g. a forgotten key, and not the optimistic "you forgot to string-convert a key")
  break
}
```

You could, of course, flip the `string` cast ok predicate back to the negative.

Also see if you think:
```
key := args[i]
i++

val := args[i]
i++
```

is any better; the biggest downside I can see is a need to `i-2` or `i-1` inside the DPanic site, depending on how you intersperse the key/val extraction with the key-string check.

In short: I bet a lot of people will end up reading this source code, since it's one step beneath a common entry point for any `godef` chasers; let's optimize its readability, not just to convince ourselves that it's correct / robust.